crates/matrix-sdk/src/error.rs

```diff
@@ -416,18 +416,13 @@ pub enum ImageError {
 /// [handling refresh tokens]: crate::ClientBuilder::handle_refresh_tokens()
 #[derive(Debug, Error, Clone)]
 pub enum RefreshTokenError {
-    /// The Matrix endpoint returned an error.
-    #[error(transparent)]
-    ClientApi(#[from] ruma::api::client::Error),
-
     /// Tried to send a refresh token request without a refresh token.
     #[error("missing refresh token")]
     RefreshTokenRequired,
 
-    /// There was an ongoing refresh token call that failed and the error could
-    /// not be forwarded.
-    #[error("the access token could not be refreshed")]
-    UnableToRefreshToken,
+    /// An error occurred interacting with the native Matrix authentication API.
+    #[error(transparent)]
+    MatrixAuth(Arc<HttpError>),
```

Comment:
So `HttpError` and `RefreshTokenError` are now mutually recursive? I really don't like that. Would it be possible to remove the refresh token variant from `HttpError`?

Reply:
Well the issue here is that `HttpError` happens when interacting with the Matrix API. That happens both when making a "regular" request, and when trying to refresh the access token.

Now we need a way to make a difference when a "regular" request fails because the access token didn't work but we tried to refresh the access token.

To not have recursive types, we would need two `HttpError` types, one with the `RefreshToken` variant and the other one without.